samples/02.branding-styling-and-customization/g.send-timeout/README.md

```diff
@@ -0,0 +1,309 @@
++     <a href="#retry=false">Show retry immediately</a>
++     <a href="#retry=1000">Show retry after 1 second</a>
++     <a href="#retry=5000">Show retry after 5 seconds</a>
++     <a href="#retry=120000">Show retry after 2 minutes</a>
++     <a href="#retry=fn">Custom retry</a>
++   </p>
+    <script>
+```
+
+Then, we will style the button bar by adding the following CSS styles.
+
+```diff
+  html,
+  body {
+    height: 100%;
+  }
+  body {
+    margin: 0;
+  }
+
+  #webchat {
+    height: 100%;
+    width: 100%;
+  }
+
++ #buttonBar {
++   display: flex;
++   flex-wrap: wrap;
++   left: 10px;
++   margin: 0;
++   position: absolute;
++   top: 10px;
++ }
++
++ #buttonBar > a {
++   background-color: White;
++   border: solid 2px #0063b1;
++   color: #0063b1;
++   font-family: Calibri, 'Helvetica Neue', Arial, sans-serif;
++   font-size: 80%;
++   margin: 0 10px 10px 0;
++   padding: 5px 8px;
++   text-decoration: none;
++ }
++
++ #buttonBar > a.selected {
++   background-color: #0063b1;
++   color: White;
++ }
+```
+
+When the page is loaded, you should see the button bar is loaded.
+
+### Hooking the button bar
+
+When clicking on the button bar, the URL will be updated with a new retry value in hash. `hashchange` event will be intercepted and update `sendTimeout` value in Web Chat by calling `renderWebChat`.
+
+First, call `renderWebChat` on initial page load and on every `hashchange` event. The `directLine` object will be memoized so it will not be recreated and reconnected on every call.
+
+```diff
+  const res = await fetch('https://webchat-mockbot.azurewebsites.net/directline/token', { method: 'POST' });
+  const { token } = await res.json();
+  const directLine = window.WebChat.createDirectLine({ token });
+
++ function renderWebChat() {
+    window.WebChat.renderWebChat(
+      {
+-       directLine: window.WebChat.createDirectLine({ token })
++       directLine
+      },
+      document.getElementById('webchat')
+    );
++ }
+
++ window.addEventListener('hashchange', renderWebChat);
+
++ renderWebChat();
+
+  document.querySelector('#webchat > *').focus();
+```
+
+Then, we will read the `sendTimeout` value from `location.hash` and update Web Chat.
+
+```diff
+  const res = await fetch('https://webchat-mockbot.azurewebsites.net/directline/token', { method: 'POST' });
+  const { token } = await res.json();
+  const directLine = window.WebChat.createDirectLine({ token });
+
++ function getHash() {
++   const hash = new URLSearchParams((location.hash || '').substr(1));
++   const hashJSON = {};
++
++   for (const [key, value] of hash.entries()) {
++     hashJSON[key] = value;
++   }
++
++   return hashJSON;
++ }
+
+  function renderWebChat() {
++   const { retry } = getHash();
++   const sendTimeout =
++     retry === 'fn'
++       ? activity => {
++           console.group('Running custom function to evaluate send timeout for activity.');
++           console.log(activity);
++           console.log('Returning 2 seconds.');
```

Comment:
?

Reply:
This is for demonstrating developers can add their own code for deciding how long an activity can wait before time out.